sotodlib/toast/ops/dark_template.py

```diff
@@ -0,0 +1,338 @@
+            # periodicity issues
+            embedded = np.zeros(2 * nsample)
+            embedded[:nsample] = template
+            lowpassed_template = scipy.signal.convolve(
+                embedded, lowpass, mode="same"
+            )
+            lowpassed_template = lowpassed_template[:nsample]
+            # Correct the ends for the zero padding (average includes zeros)
+            frac = np.ones(template.size)
+            frac[:nsum // 2] = nsum / np.arange(nsum // 2, nsum)
+            frac[-nsum // 2:] = nsum / np.arange(nsum, nsum // 2, -1)
+            lowpassed_template *= frac
+            # Regularize the tail
+            nsum_tail = nsum // 10
+            lowpassed_template[:nsum_tail] = np.mean(template[:nsum_tail])
+            lowpassed_template[-nsum_tail:] = np.mean(template[-nsum_tail:])
+            lowpassed.append(lowpassed_template)
+        lowpassed = np.vstack(lowpassed)
+        return lowpassed
+
+
+    @function_timer
+    def _load_dark_templates(self, ob):
+        """ Try loading dark bolometer templates
+        """
+        log = Logger.get()
+        comm = ob.comm.comm_group
+
+        # See if the templates were already cached
+
+        if self.cache_dir is not None:
+            fname_cache = os.path.join(
+                self.cache_dir,
+                f"dark_templates.{ob.name}.pck",
+            )
+            if os.path.isfile(fname_cache):
+                if comm is None or comm.rank == 0:
+                    with open(fname_cache, "rb") as f:
+                        templates = pickle.load(f)
+                    log.debug(f"Loaded dark templates from {fname_cache}")
+                else:
+                    templates = None
+                self._share_templates(ob, templates)
+                return ob.shared[self.key]
+
+        return None
+
+    @function_timer
+    def _share_templates(self, ob, templates):
+        """ Put the dark templates in shared memory
+
+        """
+        comm = ob.comm.comm_group
+
+        if templates is None:
+            ntemplate, nsample = None, None
+        else:
+            ntemplate, nsample = templates.shape
+
+        if comm is not None:
+            ntemplate = comm.bcast(ntemplate)
+            nsample = comm.bcast(nsample)
+
+        ob.shared.create_column(
+            self.key,
+            shape=(nsample, ntemplate),
+            dtype=np.float64,
+        )
+        ob.shared[self.key].set(templates.T, offset=(0, 0), fromrank=0)
+        return
+
+    @function_timer
+    def _save_dark_templates(self, ob, templates):
+        """ Write the dark templates in a pickle file
+
+        """
+        # See if we are caching the templates
+
+        log = Logger.get()
+        comm = ob.comm.comm_group
+
+        if self.cache_dir is None:
+            return
+
+        if comm is None or comm.rank == 0:
+            fname_cache = os.path.join(
+                self.cache_dir,
+                f"dark_templates.{ob.name}.pck",
```

Comment:
We can revisit formats and locations later.

Reply:
Indeed. This is just a placeholder for a more effective format.